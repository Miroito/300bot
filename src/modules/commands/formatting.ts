import { Character, CharacterWithOutfitWithLeader } from '@planetside/types'

const factionEmojis = [
  'Unknown Faction',
  '{emoji:faction_logo_vs|VS}',
  '{emoji:faction_logo_nc|NC}',
  '{emoji:faction_logo_tr|TR}',
  '{emoji:faction_logo_ns|NS}',
]

export const formatChacarcterWithFaction = (
  character: Character | CharacterWithOutfitWithLeader,
): string => {
  return `**${character.name.first}** ${
    factionEmojis[Number(character.factionId)]
  }${
    character.factionId === '4' &&
    'outfitMember' in character &&
    character.outfitMember
      ? ` ${
          factionEmojis[Number(character.outfitMember.outfit.leader.factionId)]
        }`
      : ''
  }`
}
