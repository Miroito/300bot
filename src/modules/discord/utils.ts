import discord from 'discord.js'
import { schedule } from '@app/utils/async'
import { log } from '@app/utils/log'
import { reviveDatabase } from '@database/revive'

export const getTextChannel = (
  client: discord.Client,
  id: string,
): discord.TextChannel | null => {
  const channel = client.channels.resolve(id)
  if (channel instanceof discord.TextChannel) {
    return channel
  }
  return null
}

export const getDMChannel = async (
  client: discord.Client,
  id: string,
): Promise<discord.DMChannel | null> => {
  const channel = await client.channels.fetch(id)
  if (channel instanceof discord.DMChannel) {
    return channel
  }
  return null
}

export const formatWithEmojis = (
  channel: discord.Channel,
  message: string,
): string => {
  // eslint-disable-next-line unicorn/no-unsafe-regex
  const emojiExpressionRegex = /{emoji:([^|]*)(?:\|([^}]*))?}/g

  if (channel instanceof discord.TextChannel) {
    return message.replace(
      emojiExpressionRegex,
      (_match, emojiName, fallbackText) => {
        const emoji = channel.guild.emojis.cache.find(
          ({ name }) => name === emojiName,
        )
        if (emoji) {
          return emoji.toString()
        } else if (typeof fallbackText === 'string') {
          return fallbackText
        } else {
          return ''
        }
      },
    )
  } else {
    return message.replace(
      emojiExpressionRegex,
      (_match, _emojiName, fallbackText) => {
        if (typeof fallbackText === 'string') {
          return fallbackText
        } else {
          return ''
        }
      },
    )
  }
}

export const removeReaction = async (
  reaction: discord.MessageReaction,
  user: discord.User,
): Promise<void> => {
  try {
    await reaction.users.remove(user)
  } catch (error) {
    log.error('Error removing reaction:', error)
  }
}

export const reviveMember = async (
  member: discord.GuildMember,
  deadRole: discord.Role,
): Promise<void> => {
  const guildId = member.guild.id
  const userId = member.id

  await member.roles.remove(deadRole)
  await reviveDatabase.delete(`${guildId}.${userId}`)
}

export const killMember = async (
  member: discord.GuildMember,
  deadRole: discord.Role,
  reviveDelayMs: number,
): Promise<void> => {
  const userId = member.id
  const guildId = member.guild.id

  const reviveTime = Date.now() + reviveDelayMs

  await reviveDatabase.set(`${guildId}.${userId}`, reviveTime)
  await member.roles.add(deadRole)

  void schedule(async () => {
    try {
      await reviveMember(member, deadRole)
    } catch (e) {
      log.error('Cannot revive member:', e)
    }
  }, reviveDelayMs)
}
