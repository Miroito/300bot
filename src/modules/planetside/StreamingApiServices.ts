import { getUTCShort } from '@app/utils/time'
import { dmTrackerDatabase } from '@database/dmtracker'
import { client } from '@discord/client'
import { formatWithEmojis, getDMChannel, getTextChannel } from '@discord/utils'
import { censusApi } from '@planetside/CensusApi'
import { streamingApi } from '@planetside/StreamingApi'
import {
  getAlertType,
  getAlertState,
  getContinentName,
} from '@planetside/resources'
import { alertTrackerDatabase } from '../database/alerttracker'

export const startTrackingService = (): void => {
  const sendCharacterStatus = async (characterId: string, online: boolean) => {
    const channelIds = dmTrackerDatabase.get(characterId)
    if (!channelIds) return
    const character = await censusApi.getCharacterName({ characterId })
    if (!character) return
    const characterName = character.name.first
    const message = online
      ? `**${characterName}** is online!`
      : `**${characterName}** is offline.`
    Object.keys(channelIds).forEach((channelId) => {
      void (async () => {
        const channel = await getDMChannel(client, channelId)
        if (channel) {
          void channel.send(message)
        }
      })()
    })
  }

  streamingApi.on('playerLogin', async ({ characterId }) => {
    await sendCharacterStatus(characterId, true)
  })
  streamingApi.on('playerLogout', async ({ characterId }) => {
    await sendCharacterStatus(characterId, false)
  })
}

export const startAlertTrackerService = (): void => {
  const sendAlertMessage = (message: string) => {
    message = `[${getUTCShort()}] ${message}`
    const channelIds = Object.values(alertTrackerDatabase.root)
    channelIds.forEach((channelId) => {
      const channel = getTextChannel(client, channelId)
      if (channel) {
        void channel.send(formatWithEmojis(channel, message))
      }
    })
  }

  streamingApi.on('continentLock', ({ zoneId }) => {
    const continentName = getContinentName(zoneId)
    if (!continentName) return
    sendAlertMessage(`:lock: **${continentName}** has been locked.`)
  })
  streamingApi.on('continentUnlock', ({ zoneId }) => {
    const continentName = getContinentName(zoneId)
    if (!continentName) return
    sendAlertMessage(`:unlock: **${continentName}** has been unlocked.`)
  })
  streamingApi.on('metagameEvent', (event) => {
    const alertType = getAlertType(event.metagameEventId)
    const alertState = getAlertState(event.metagameEventState)
    if (!alertType || !alertState) return
    const [continentName, factionAlias] = alertType

    const factionData = [
      {
        alias: 'TR',
        emoji: '{emoji:faction_logo_tr}',
        emojiWithFallback: '{emoji:faction_logo_tr|**TR**}',
        result: Math.floor(Number(event.factionTr)),
      },
      {
        alias: 'NC',
        emoji: '{emoji:faction_logo_nc}',
        emojiWithFallback: '{emoji:faction_logo_nc|**NC**}',
        result: Math.floor(Number(event.factionNc)),
      },
      {
        alias: 'VS',
        emoji: '{emoji:faction_logo_vs}',
        emojiWithFallback: '{emoji:faction_logo_vs|**VS**}',
        result: Math.floor(Number(event.factionVs)),
      },
    ]

    if (alertState === 'started') {
      const faction = factionData.find((item) => item.alias === factionAlias)
      if (!faction) return
      sendAlertMessage(
        `${faction.emoji} **${faction.alias}** have started an alert on **${continentName}** with **${faction.result}%** territory control.`,
      )
    } else {
      const factionResults = factionData
        .map(
          ({ emojiWithFallback, result }) =>
            `${emojiWithFallback} **${result}%**`,
        )
        .join(' ')

      sendAlertMessage(
        `:rotating_light: The alert on **${continentName}** has ended: ${factionResults}`,
      )
    }
  })
}
