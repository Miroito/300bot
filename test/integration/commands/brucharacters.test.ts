import brucharacters from '@app/modules/commands/common/brucharacters'
import { getCommandRunner } from '@test/utils/commands'
import { mockCensusApi } from '@test/utils/planetside'

describe('brucharacters', () => {
  const runCommand = getCommandRunner(brucharacters)

  it("should say that the list of bru's characters is empty", async () => {
    const reply = await runCommand('+brucharacters list')
    expect(reply).toEqual("The list of Bru's characters is empty.")
  })

  it('should add and remove characters from the database regardless of case', async () => {
    const requestData = mockCensusApi({
      character_name_list: [
        {
          characterId: 'testaddremove',
          name: { first: 'CustomChar', firstLower: 'customchar' },
        },
      ],
    })

    const reply = await runCommand('+brucharacters add CUStomChar')
    expect(requestData).toMatchSnapshot()
    expect(reply).toEqual(
      "**CustomChar** has been added to the list of Bru's characters.",
    )

    const requestData1 = mockCensusApi({
      character_name_list: [
        {
          characterId: 'testaddremove',
          name: { first: 'CustomChar', firstLower: 'customchar' },
        },
      ],
    })

    const reply1 = await runCommand('+brucharacters remove CUStomChar')
    expect(requestData1).toMatchSnapshot()
    expect(reply1).toEqual(
      "**CustomChar** has been removed from the list of Bru's characters.",
    )
  })

  it('should not remove a character that is not in the list', async () => {
    const requestData = mockCensusApi({
      character_name_list: [
        {
          characterId: 'testunknown',
          name: { first: 'unknownChar', firstLower: 'unknownchar' },
        },
      ],
    })

    const reply = await runCommand('+brucharacters remove unknownChar')
    expect(requestData).toMatchSnapshot()
    expect(reply).toEqual(
      "**unknownChar** is not listed as one of Bru's characters.",
    )
  })
})
