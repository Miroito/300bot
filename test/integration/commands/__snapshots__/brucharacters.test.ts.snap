// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`brucharacters should add and remove characters from the database regardless of case 1`] = `
Object {
  "pathname": "/s:test/get/ps2:v2/character_name",
  "searchParams": Object {
    "name.first_lower": "customchar",
  },
}
`;

exports[`brucharacters should add and remove characters from the database regardless of case 2`] = `
Object {
  "pathname": "/s:test/get/ps2:v2/character_name",
  "searchParams": Object {
    "name.first_lower": "customchar",
  },
}
`;

exports[`brucharacters should display the list of characters in the database 1`] = `
Object {
  "pathname": "/s:test/get/ps2:v2/character_name",
  "searchParams": Object {
    "name.first_lower": "customcharlist",
  },
}
`;

exports[`brucharacters should display the list of characters in the database 2`] = `
Object {
  "pathname": "/s:test/get/ps2:v2/character_name",
  "searchParams": Object {
    "c:limit": "100",
    "character_id": "testlist",
  },
}
`;

exports[`brucharacters should not remove a character that is not in the list 1`] = `
Object {
  "pathname": "/s:test/get/ps2:v2/character_name",
  "searchParams": Object {
    "name.first_lower": "unknownchar",
  },
}
`;

exports[`brucharacters should say that the character is already in the list 1`] = `
Object {
  "pathname": "/s:test/get/ps2:v2/character_name",
  "searchParams": Object {
    "name.first_lower": "customchardup",
  },
}
`;

exports[`brucharacters should say that the character is already in the list 2`] = `
Object {
  "pathname": "/s:test/get/ps2:v2/character_name",
  "searchParams": Object {
    "name.first_lower": "customchardup",
  },
}
`;
